### PR TITLE
documents: optimize availability

### PR DESCRIPTION
**Pull request by jma**
- Uses Elasticsearch to compute the documents, items and holdings availability.
- Search only one organisation when an organisation is retrived using a view code.

## Why are you opening this PR?

- Which task/US does it implement?
- Which issue does it fix?

## Dependencies

My PR depends on the following `rero-ils-ui`'s PR(s):

* rero/rero-ils-ui#<xx>

## How to test?

- What command should I have to run to test your PR?
- What should I test through the UI?

### CODE REVIEW

File: rero_ils/modules/items/api/circulation.py [hunk 1]
```diff
@@ -1276,26 +1277,19 @@ def get_loan_states_for_an_item(self):
...
-            return False
-        if self.temp_item_type_negative_availability:
-            return False
-        if self.is_issue and self.issue_status != ItemIssueStatus.RECEIVED:
-            return False
-        if self.item_has_active_loan_or_request() > 0:
+        items_query = ItemsSearch().available_query()
+
+        # check item availability
+        if not bool(items_query.filter(Q('term', pid=self.pid)).count()):
```
Comment (rerowep) [inline rero_ils/modules/items/api/circulation.py:1287]: Why we need `Q()` in this filter?
Could we write:
```python
if items_query.filter('term', pid=self.pid).count() == 0:
    return False
```

**Review by rerowep — CHANGES_REQUESTED** [review]

File: rero_ils/modules/documents/jsonschemas/documents/document_agent_local-v0.0.1.json [hunk 2]
```diff
@@ -0,0 +1,59 @@
+{
```
[the remark this answers is not in the document]
Comment (jma) [reply rero_ils/modules/documents/jsonschemas/documents/document_agent_local-v0.0.1.json:1]: It is a mistake, sorry.

File: rero_ils/modules/holdings/api.py [hunk 3]
```diff
@@ -260,14 +268,61 @@ def vendor(self):
...
+
+        return [
+            hit.item_pid.value for hit in loan_query.source('item_pid').scan()
+        ]
+
+    def is_available(self):
+        """Get availability for the current holding."""
+        # -------------- Holdings --------------------
+        # unavailable if masked
+        if self.get('_masked', False):
```
[the remark this answers is not in the document]
Comment (jma) [reply rero_ils/modules/holdings/api.py:304]: I know but as masked is a boolean, I think it is more readable.

File: rero_ils/modules/documents/api.py [hunk 4]
```diff
@@ -127,27 +127,122 @@ def _validate(self, **kwargs):
...
+        :param org_pid: str - the organisation pid.
+        :returns: int, int - the number of available and electronic holdings.
+        """
+        from rero_ils.modules.holdings.api import HoldingsSearch
+
+        # create the holding search query
+        holding_query = HoldingsSearch().available_query()
+
+        # filter by the current document
+        filters = Q('term', document__pid=pid)\
```
Comment (Garfield-fr) [inline rero_ils/modules/documents/api.py:143]: Remove `\`

File: rero_ils/modules/loans/api.py [hunk 5]
```diff
@@ -72,6 +72,12 @@ class Meta:
 
         default_filter = None
 
+    def unavailable_query(self):
+        """."""
```
Comment (Garfield-fr) [inline rero_ils/modules/loans/api.py:76]: Missing docstring

**Review by Garfield-fr — APPROVED** [review]

**Review by rerowep — APPROVED** [review]

File: rero_ils/modules/documents/api.py [hunk 6]
```diff
@@ -127,27 +127,125 @@ def _validate(self, **kwargs):
...
+        # unavailable if no holdings exists
+        if not n_available_holdings:
+            return False
+
+        # -------------- Items --------------------
+        # get the available item pids
+        available_item_pids = cls.get_available_item_pids(pid, org_pid)
+
+        # unavailable if no items exists
+        if not available_item_pids:
```
Comment (zannkukai) [inline rero_ils/modules/documents/api.py:239]: cosmetic : could be 
```python
if not (available_items_pids := cls.get_available_item_pids(pid, org_pid)):
    ...
```
But maybe less readable... Myself, i'm not sure

File: rero_ils/modules/holdings/api.py [hunk 7]
```diff
@@ -260,14 +268,65 @@ def vendor(self):
...
+
+        :param item_pids: [str] - the list of the item pids.
+        :returns: the list of the item pids having active loans.
+        """
+        from rero_ils.modules.loans.api import LoansSearch
+
+        loan_query = LoansSearch().unavailable_query()
+
+        # the loans corresponding to the given item pids
+        loan_query = loan_query.filter(Q('terms', item_pid__value=item_pids))
```
Comment (zannkukai) [inline rero_ils/modules/holdings/api.py:294]: Why not using same notation than everywhere else ? `filters = Q('terms, ...)`

File: rero_ils/modules/items/api/api.py [hunk 8]
```diff
@@ -62,6 +64,37 @@ class Meta:
...
+            .scan()
+        ]
+        if not_available_item_types:
+            # negative availability item type and not temporary item types
+            has_items_filters = \
+                Q('terms', item_type__pid=not_available_item_types)
+            has_items_filters &= ~Q('exists', field='temporary_item_type')
+            # temporary item types with negative availability
+            has_items_filters |= Q(
+                'terms', temporary_item_type__pid=not_available_item_types)
```
Comment (zannkukai) [inline rero_ils/modules/items/api/api.py:93]: cosmetic : could be simplify

```python
has_items_filters =
    Q('terms', item_type__pid=not_available_item_types)
    & ~Q('exists', field='temporary_item_type')
    | Q('terms', temporary_item_type__pid=not_available_item_types)
```

**Review by zannkukai — APPROVED** [review]

**Pull request merged**